File: Python/lab19/Consultant.py
from Observer import Observer


class Consultant(Observer):
    consultants = []

    def __init__(self, full_name, rang, is_active):
        self.full_name = full_name
        self.rang = rang
        self.is_active = is_active
        Consultant.consultants.append(self)

    def remove(self):
        Consultant.consultants.remove(self)
        self.notify(self, "remove")

    def add(self, consultant):
        self.notify(consultant, "add")

    def change_status(self):
        if self.is_active:
            self.is_active = False
            self.notify(self, "false")
        else:
            self.is_active = True
            self.notify(self, "true")

    def notify(self, consultant, command):
        for c in Consultant.consultants:
            if c != consultant:
                if command == "remove":
                    print(f"Consultant: {c.full_name} get that consultant {consultant.full_name} is dismissed")
                elif command == "add":
                    print(f"Consultant: {c.full_name} get that consultant {consultant.full_name} is hiring")
                elif command == "true":
                    print(f"Consultant: {c.full_name} get that consultant {consultant.full_name} took over the shift")
                elif command == "false":
                    print(f"Consultant: {c.full_name} get that consultant {consultant.full_name} finished the shift")

File: Java/lab19/src/main/java/org/example/Consultant.java
package org.example;

import java.util.ArrayList;
import java.util.List;

public class Consultant implements Observer{
    public static List<Consultant> consultants=new ArrayList<>();
    private final String fullName;
    private String rang;
    private Boolean isActive;

    public Consultant(String fullName, String rang, Boolean isActive) {
        this.fullName = fullName;
        this.rang = rang;
        this.isActive = isActive;
        consultants.add(this);
    }

    @Override
    public void remove() {
        consultants.remove(this);
        notufy(this,"remove");
    }

    @Override
    public void add(Consultant consultant) {
        notufy(consultant,"add");
    }
    @Override
    public void changeStatus() {
        if(isActive){
            isActive=false;
            notufy(this,"false");
        }else {
            isActive=true;
            notufy(this,"true");
        }
    }
    @Override
    public void notufy(Consultant consultant,String command) {
        for(Consultant c:consultants){
            if(!c.equals(consultant)){
                switch (command) {
                    case "remove" -> System.out.println("Consultant: " + c.fullName
                            + " get, that consultant " + consultant.fullName + " is dismissed");
                case "add" -> System.out.println("Consultant: " + c.fullName
                        + " get, that consultant " + consultant.fullName + " is hiring");
                    case "true" -> System.out.println("Consultant: " + c.fullName
                            + " get, that consultant " + consultant.fullName + " took over the shift");
                    case "false" -> System.out.println("Consultant: " + c.fullName
                            + " get, that consultant " + consultant.fullName + " finished the shift");
                }
            }

        }
    }
}
